src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ServiceCategoryService.kt

```diff
@@ -11,4 +13,25 @@ import java.util.UUID
 @Transactional
 class ServiceCategoryService(val repository: ServiceCategoryRepository) {
   fun getServiceCategoryByID(id: UUID): ServiceCategory? = repository.findByIdOrNull(id)
+  fun getServiceCategoryByIDAndContractReference(id: UUID, contractReference: String): ServiceCategory? {
+    val serviceCategoryWithAllDesiredOutcomes = repository.findByIdOrNull(id)
+    val serviceCategoryWithFilteredOutcomes = serviceCategoryWithAllDesiredOutcomes?.copy()
+    serviceCategoryWithFilteredOutcomes?.desiredOutcomes = serviceCategoryWithFilteredOutcomes?.desiredOutcomes?.filter { isDesiredOutcomeValidForContract(it, contractReference) }.orEmpty()
+
+    return serviceCategoryWithFilteredOutcomes
+  }
+
+  fun isDesiredOutcomeValidForContract(desiredOutcome: DesiredOutcome, contractReference: String): Boolean {
+    if (desiredOutcome.deprecatedAt != null) {
+      return false
+    }
+
+    desiredOutcome.desiredOutcomeFilterRules.forEach {
+      return when (it.ruleType) {
+        RuleType.INCLUDE -> it.matchData.contains(contractReference)
+        RuleType.EXCLUDE -> !it.matchData.contains(contractReference)
+      }
+    }
+    return true
```

Comment:
```suggestion
  val foundItem =
      desiredOutcome.desiredOutcomeFilterRules.find { it.matchData.contains(contractReference) } ?: return true
    return when (foundItem.ruleType) {
      RuleType.INCLUDE -> foundItem.matchData.contains(contractReference)
      RuleType.EXCLUDE -> !foundItem.matchData.contains(contractReference)
    }
```
another little interesting thing just to show a different way of doing it if you wanted.

Reply:
I'd perhaps have to test this. But we also want to apply each rule even when a match against contract wasn't found. For example an 'INCLUDE' rule should be remove the DO when it doesn't match.